Refactor element access of pointer arrays into a separate function

antlechner opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->
~This PR is based on https://github.com/diffblue/cbmc/pull/4524. Please only review the last commit.~

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

antlechner committed:
Refactor array access into a separate function

This does not save many lines of code, but the way we access arrays
using pointer arithmetic can be a bit confusing and it makes sense to
abstract it away and keep the documentation of why we do things this way
in one place.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4b7cf97f30be8c030021b69a6e598f6e1098f40a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108164284

smowton commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -191,6 +191,12 @@ dereference_exprt checked_dereference(const exprt &expr)
 }
 
+dereference_exprt
+element_at_pointer_array_index(const exprt &pointer, const exprt &index)
```
Suggested name `array_element_from_pointer`

antlechner replied:
Changed the name to your suggestion.

smowton reviewed (commented):
This is not JBMC-specific -- suggest move to `util/`

owen-mc-diffblue commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -191,6 +191,12 @@ dereference_exprt checked_dereference(const exprt &expr)
+dereference_exprt
+element_at_pointer_array_index(const exprt &pointer, const exprt &index)
+{
+  return dereference_exprt{plus_exprt{pointer, index}};
```
:bulb: Would it be a good idea to have something like `INVARIANT(can_cast_type<pointer_typet>(pointer.type()));` at the beginning of this function?

antlechner replied:
Done (using `PRECONDITION`).

antlechner commented:
This is rebased on develop now.
@smowton I moved it to a new file `util/array_element_from_pointer.h` and changed the documentation slightly. I don't actually know how we represent arrays for C / when we represent them as pointers to the first element, so let me know if anything should be added to the documentation as it is now.

tautschnig commented on src/util/array_element_from_pointer.h:
```diff
@@ -0,0 +1,32 @@
+array_element_from_pointer(const exprt &pointer, const exprt &index)
+{
+  PRECONDITION(can_cast_type<pointer_typet>(pointer.type()));
+  return dereference_exprt{plus_exprt{pointer, index}};
```
While at it, maybe also do `PRECONDITION(index.type().id() == ID_signedbv || index.type().id() == ID_unsignedbv);`? I think those are the only two reasonable options, although maybe we could also work with some other bitvector types.

antlechner replied:
I added the precondition. If we need other types in the future we can always add them later.

tautschnig reviewed (approved):
I think this is fine for C as well. I'm wondering, however, whether `util/pointer_offset_sum.h` would also be an ok header to place this? No particularly strong feelings about it, though.

smowton commented on src/util/array_element_from_pointer.h:
```diff
@@ -0,0 +1,32 @@
+/*******************************************************************\
+
+Module: Util
```
This is usually a description of the file, not the directory

antlechner replied:
I had seen `Module: Util` in another file and just copied that. Most files in `util` seem to just leave this empty... Anyway, I changed this one to "Element access in a pointer array".

smowton commented on src/util/array_element_from_pointer.h:
```diff
@@ -0,0 +1,32 @@
+/// given index of a pointer array:
+/// `*(pointer + index)`
+/// Arrays are sometimes (always in JBMC) represented as a pointer to their
+/// first element. This is because we cannot know the type of an array of
```
Suggest replace the second sentence with just `, especially when their length is uncertain`

antlechner replied:
I replaced a part of the second sentence with that. :slightly_smiling_face:

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 779f5dfb0e8ce4e44c451b59ebfca82b2e441009).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108340077

antlechner commented:
@tautschnig `pointer_offset_sum.h` contains only one function with the same name as the file, so I'll keep this one separate for now to avoid confusion.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2088544ca4d07f73fbf60ab2b42f949f9bfac513).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108571279
